pkg/codegen/dotnet/gen.go

```diff
@@ -752,7 +752,7 @@ func (pt *plainType) genOutputType(w io.Writer, level int) {
 	// Generate the constructor body.
 	fmt.Fprintf(w, "%s    {\n", indent)
 	for _, prop := range pt.properties {
-		paramName := csharpIdentifier(prop.Name)
+		paramName := csharpIdentifier(LowerCamelCase(pt.mod.propertyName(prop)))
```

Comment:
because it is a parameter name for the constructor

Reply:
Ah right, I think I was thrown by the example code having ` public Input<string>? myBar { get; set; }` because that's an odd C# name but thats just because the schema has the literal "myBar" not because of these lowerCamelCase calls.
Wonder if it's worth adding another PascalCase field to show that these calls _do_ matter then.

Reply:
OK so I've added another field `input_data` which has a csharp language override of `InputData`. This makes the constructor parameter name `inputData` which is what we want. 

However, I am a bit concerned here about how this type is actually deserialized at runtime 🤔 because the fields are _not_ annotated with their original field name (e.g. `input_data`) and the only information available about the original field name is in the constructor parameter name. Now we changed `input_data` to `inputData`, will the serializer know how to fill in the constructor at runtime?

Reply:
I think if you call the field inputData and the CSharp name something like "MyInputData" it would probably all work out? We don't generally expect snake_case names in schemas.

Reply:
> However, I am a bit concerned here about how this type is actually deserialized at runtime 🤔 because the fields are not annotated with their original field name (e.g. input_data) and the only information available about the original field name is in the constructor parameter name. Now we changed input_data to inputData, will the serializer know how to fill in the constructor at runtime?

This has actually been a concern for me as well in a provider I am building.

> I think if you call the field inputData and the CSharp name something like "MyInputData" it would probably all work out? We don't generally expect snake_case names in schemas.

What if the schema has snake_case property names because the provider API uses that case for requests?

Reply:
@Frassle I've been looking into the [deserialization logic](https://github.com/pulumi/pulumi-dotnet/blob/main/sdk/Pulumi/Serialization/Converter.cs#L188-L212) and it seems that the names of the parameters _must_ match the fields of input data dictionary at runtime:
```cs
var constructor = GetPropertyConstructor(targetType);
var (dictionary, tempException) = TryEnsureType<ImmutableDictionary<string, object>>(context, val);
var constructorParameters = constructor.GetParameters();
var arguments = new object?[constructorParameters.Length];
for (int i = 0, n = constructorParameters.Length; i < n; i++)
{
    var parameter = constructorParameters[i];
    dictionary!.TryGetValue(parameter.Name!, out var argValue);
    arguments[i] = ConvertObject(warn, "...", argValue, parameter.ParameterType);
}
```
Which means this PR will cause the parameters to use default values (due to `TryGetValue` failing) and this is definitely not what we want.

Reply:
Ah yet again another codegen logic not being that clear without runtime tests issue.

At any rate I thought this was what the Pulumi.InputAttributes were for? They record the original pulumi name.

Reply:
Apparently input attributes only work for classes inheriting from `InputArgs` such as `ResourceArgs` so generated input types and output types don't have this functionality 😓

Reply:
eek ok that probably needs fixing

Reply:
> Slightly unrelated but is this sort of translation handled in the other SDKs, except for Python in this case due to snake_case being the canonical casing there anyway.

It should be handled in all the SDKs, they all differ slightly in what a schema name is vs what a language name is.

Reply:
Sorry, I meant the marshalling/unmarshalling of properties that have overrides when the properties in the schema use snake_case. Specifically, the inputs map delivered to the provider -- would that have the original property names as defined by the schema?

Reply:
Yes every SDK should be handling that _somehow_. The inputs sent to the provider should always use the keys from the schema.